packages/block-editor/src/components/provider/index.js

```diff
@@ -0,0 +1,152 @@
+
+/**
+ * Higher-order component which renders the original component with the current
+ * registry context passed as its `registry` prop.
+ *
+ * @param {WPComponent} OriginalComponent Original component.
+ *
+ * @return {WPComponent} Enhanced component.
+ */
+const withRegistry = createHigherOrderComponent(
+	( OriginalComponent ) => ( props ) => (
+		<RegistryConsumer>
+			{ ( registry ) => (
+				<OriginalComponent
+					{ ...props }
+					registry={ registry }
+				/>
+			) }
+		</RegistryConsumer>
+	),
+	'withRegistry'
+);
+
+class BlockEditorProvider extends Component {
+	componentDidMount() {
+		this.props.updateEditorSettings( this.props.settings );
+		this.props.resetBlocks( this.props.value );
+		this.attachChangeObserver( this.props.registry );
+	}
+
+	componentDidUpdate( prevProps ) {
+		const {
+			settings,
+			updateEditorSettings,
+			value,
+			resetBlocks,
+			registry,
+		} = this.props;
+
+		if ( settings !== prevProps.settings ) {
+			updateEditorSettings( settings );
+		}
+
+		if ( registry !== prevProps.registry ) {
+			this.attachChangeObserver( registry );
+		}
+
+		if ( this.isSyncingOutcomingValue ) {
+			this.isSyncingOutcomingValue = false;
+		} else if ( value !== prevProps.value ) {
+			this.isSyncingIncomingValue = true;
+			resetBlocks( value );
+		}
+	}
+
+	componentWillUnmount() {
+		if ( this.unsubscribe ) {
+			this.unsubscribe();
+		}
+	}
+
+	/**
+	 * Given a registry object, overrides the default dispatch behavior for the
+	 * `core/block-editor` store to interpret a state change and decide whether
+	 * we should call `onChange` or `onInput` depending on whether the change
+	 * is persistent or not.
+	 *
+	 * This needs to be done synchronously after state changes (instead of using
+	 * `componentDidUpdate`) in order to avoid batching these changes.
+	 *
+	 * @param {WPDataRegistry} registry     Registry from which block editor
+	 *                                      dispatch is to be overriden.
+	 */
+	attachChangeObserver( registry ) {
+		if ( this.unsubscribe ) {
+			this.unsubscribe();
+		}
+
+		const {
+			getBlocks,
+			isLastBlockChangePersistent,
+		} = registry.select( 'core/block-editor' );
+
+		let blocks = getBlocks();
+		let isPersistent = isLastBlockChangePersistent();
+
+		this.unsubscribe = registry.subscribe( () => {
+			const {
+				onChange,
+				onInput,
+			} = this.props;
+			const newBlocks = getBlocks();
+			const newIsPersistent = isLastBlockChangePersistent();
+			if ( newBlocks !== blocks && this.isSyncingIncomingValue ) {
+				this.isSyncingIncomingValue = false;
+				blocks = newBlocks;
+				isPersistent = newIsPersistent;
+				return;
+			}
+
+			if (
+				newBlocks !== blocks ||
+				// This happens when a previous input is explicitely marked as persistent.
+				( newIsPersistent && ! isPersistent )
+			) {
+				blocks = newBlocks;
+				isPersistent = newIsPersistent;
+
+				this.isSyncingOutcomingValue = true;
```

Comment:
Something I was starting to not like about my previous implementation was having a hard assumption that the parent component was rendering `BlockEditor` as controlled, i.e. assuming that `onInput` would cause the next rendered `value` to be the one we pass as `blocks` here. It seems like something which should regularly be true, but the lack of guarantee makes it feel a bit fragile.

Reply:
I agree with you and this is what was causing the e2e tests without the last commit but it was specific to the initial call. 

If we want it to be less fragile, we'd have to call `getBlocks` right after the `resetBlocks` call (I was doing this initially in the PR)